enhancements/machine-config/on-cluster-builds.md

```diff
@@ -0,0 +1,1009 @@
+    ReleaseOSImageChange--> |OpenShift upgrade changed \n the OS base image| IncomingConfigChange
+    IncomingConfigChange-->isOSContentChange
+    NodeController--> |desiredImage and/or desiredConfig\n annotations updated on node object| MCD
+    BuildController--> |Final image pullspec added\n to MachineConfigPool| NodeController
+    RenderController--> |Rendered MachineConfig created\n and added to MachineConfigPool| NodeController
+    isOSContentChange--> |Yes| BuildController
+    isOSContentChange--> |No| RenderController
+    ImageRegistry--> |Base OS Image Pull| ImageBuilder
+    ImageBuilder--> |Final OS Image Push| ImageRegistry
+    ImageBuilder--> |Fully-Qualified Image Pullspec| BuildController
+    ImageRegistry--> |Final OS Image Pull| bootc
+```
+
+With both modes of operation, there will be certain configurations which
+change more frequently than others either in response to other OpenShift
+operators (such as the Node Tuning Operator) or due to internal cluster
+operations (certificate rotations, global pull secret changes, etc.).
+These configurations should be only set up using bootc’s ConfigMap
+capability since they can be easier to manage via this path.
+
+The tension between these two modes is very similar (if not the same) to the
+container vs. configs paradigm. In other words, if one containerizes their
+application, should they include all of the configs within the application
+container? Or should they instead configure their application via ConfigMaps and
+/ or Secrets which are injected at runtime?
+
+#### NodeController
+
+The NodeController will be modified to watch the MachineConfigPool for
+the current container image spec. Whenever a new container image is
+available, it will update the `desiredImage` annotation on each node in
+the given pool, according to its current rules for deploying
+MachineConfigs.
+
+#### BuildController
+
+A new MCO component called BuildController will manage the lifecycle of image
+builds for custom OS content. It listens for changes to the MachineConfigPool
+and responds by starting a build whenever needed.
+
+To handle multi-arch cases, it will need to schedule more than one build for
+each architecture present within the cluster, doing each build natively on the
+target architectures. In other words, if one has both ARM64 and AMD64 worker
+nodes, a build will be scheduled on both an ARM64 node as well as an AMD64 node.
+Doing this allows us to execute the builds with as few permissions as possible.
+
+BuildController should be as modular as possible with how it performs the build
+to allow it to efficiently operate in as many different environments as
+possible. For example, if the OpenShift Image Builds (BuildsV1) facility is
+present, it could leverage that functionality directly. In absence of that
+facility, it could create its own custom build pod where it uses Buildah to
```

Comment:
If we are to not rely on availability of BuildsV1 might as well just use custom build pods as the implementation in call cases.